Fix full_like decomposition to preserve strides

masnesral opened the pull request:
Summary:
See original PR at: https://github.com/pytorch/pytorch/pull/144765, which landed internally but was reverted due to test failures. Addressing reviewer comments and trying again.

Rollback Plan:

Differential hack Revision: D78783627


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

masnesral commented:
@eellison @isuruf this should be a dup of https://github.com/pytorch/pytorch/pull/144765; having a hard time getting past various CI issues and was advised to start with a new internal diff and export it to OSS. So here it is

eellison reviewed (commented):
Can we fix the prims.full ? See comment [here](https://github.com/pytorch/pytorch/pull/144765#discussion_r2127356609) for more context.

masnesral commented on torch/_inductor/lowering.py:
```diff
@@ -6143,6 +6142,17 @@ def fill_(x, fill_value):
+        inner_fn=lambda _: ops.constant(fill_value, dtype),
+        ranges=list(x.get_size()),
+    )
+
```
@eellison, actually the test failures on the latest rev kinda look like a circular dependence @isuruf described in that comment, e.g.,
```
  File "/data/users/slarsen/pytorch-masnesral/torch/_refs/__init__.py", line 5598, in full
    return aten.full(e, fill_value)
  File "/data/users/slarsen/pytorch-masnesral/torch/_ops.py", line 1254, in __call__
    return self._op(*args, **kwargs)
  File "/data/users/slarsen/pytorch-masnesral/torch/_prims/context.py", line 153, in __torch_function__
    return func(*args, **kwargs)
  File "/data/users/slarsen/pytorch-masnesral/torch/_prims_common/wrappers.py", line 309, in _fn
    result = fn(*args, **kwargs)
  File "/data/users/slarsen/pytorch-masnesral/torch/_refs/__init__.py", line 5590, in full
    e = empty(
  File "/data/users/slarsen/pytorch-masnesral/torch/_prims_common/wrappers.py", line 309, in _fn
    result = fn(*args, **kwargs)
  File "/data/users/slarsen/pytorch-masnesral/torch/_refs/__init__.py", line 4800, in empty
    strides = utils.make_contiguous_strides_for(shape)
  File "/data/users/slarsen/pytorch-masnesral/torch/_prims_common/__init__.py", line 1711, in make_contiguous_strides_for
    multiplier *= l if is_nested_int(l) else sym_max(l, 1)  # type:ignore[assignment]
  File "/data/users/slarsen/pytorch-masnesral/torch/__init__.py", line 864, in sym_max
    return overrides.handle_torch_function(sym_max, (a, b), a, b)
  File "/data/users/slarsen/pytorch-masnesral/torch/overrides.py", line 1726, in handle_torch_function
    result = mode.__torch_function__(public_api, types, args, kwargs)
  File "/data/users/slarsen/pytorch-masnesral/torch/_prims/context.py", line 128, in __torch_function__
    return orig_func(*args, **kwargs)
  File "/data/users/slarsen/pytorch-masnesral/torch/__init__.py", line 864, in sym_max
    return overrides.handle_torch_function(sym_max, (a, b), a, b)
  File "/data/users/slarsen/pytorch-masnesral/torch/overrides.py", line 1726, in handle_torch_function
    result = mode.__torch_function__(public_api, types, args, kwargs)
  File "/data/users/slarsen/pytorch-masnesral/torch/testing/_internal/common_utils.py", line 1565, in __torch_function__
    r = func(*args, **kwargs)
  File "/data/users/slarsen/pytorch-masnesral/torch/__init__.py", line 875, in sym_max
    assert isinstance(a, all_types), type(a)
AssertionError: <class 'torch.Tensor'>
```

eellison replied:
What is the circular dependence ? We don't need prims.fill.

Just updating the existing aten.full_like decomp to use same logic that has been added to `torch/_refs/__init__.py` new decomposition would be sufficient.

amjames commented:
@eellison @masnesral FYI @isuruf is out, but he returns tomorrow. I will make sure he sees this first thing.

masnesral commented:
> FYI @isuruf is out, but he returns tomorrow. I will make sure he sees this first thing.

@isuruf it's been a comedy of challenges getting this merged internally. I'm kinda desperate to get it done. Assuming I can actually make that happen today, gonna land as is. But I'm more than happy to follow-up with any changes if you find suggestions.

masnesral committed:
Fix full_like decomposition to preserve strides (#158898)

Summary:

See original PR at: https://github.com/pytorch/pytorch/pull/144765, which landed internally but was reverted due to test failures. Addressing reviewer comments and trying again.

Test Plan:
OSS and internal CI

Rollback Plan:

Reviewed By: eellison

Differential Revision: D78783627

ZainRizvi commented:
@pytorchbot merge -f "internal diff has already landed"

Skylion007 commented on torch/_inductor/decomposition.py:
```diff
@@ -637,14 +652,36 @@ def full_like(
+        )
+        if permutation == list(range(len(permutation))):
+            return result
+        return result.permute(permutation).clone()
```
Why is this clone necessary?
